packages/graphql-codegen-cli/src/utils/watcher.ts

```diff
@@ -23,47 +22,24 @@ function emitWatching(watchDir: string) {
 }
 
 export const createWatcher = (
-  initalContext: CodegenContext,
+  initialContext: CodegenContext,
   onNext: (result: Types.FileOutput[]) => Promise<Types.FileOutput[]>
 ): Promise<void> => {
   debugLog(`[Watcher] Starting watcher...`);
-  let config: Types.Config & { configFilePath?: string } = initalContext.getConfig();
-  const files: string[] = [initalContext.filepath].filter(a => a);
-  const documents = normalizeInstanceOrArray<Types.OperationDocument>(config.documents);
-  const schemas = normalizeInstanceOrArray<Types.Schema>(config.schema);
-
-  // Add schemas and documents from "generates"
-  for (const conf of Object.keys(config.generates).map(filename => normalizeOutputParam(config.generates[filename]))) {
-    schemas.push(...normalizeInstanceOrArray<Types.Schema>(conf.schema));
-    documents.push(...normalizeInstanceOrArray<Types.OperationDocument>(conf.documents));
-    files.push(...normalizeInstanceOrArray(conf.watchPattern));
-  }
-
-  if (documents) {
-    for (const doc of documents) {
-      if (typeof doc === 'string') {
-        files.push(doc);
-      } else {
-        files.push(...Object.keys(doc));
-      }
-    }
-  }
+  let config: Types.Config & { configFilePath?: string } = initialContext.getConfig();
 
-  for (const s of schemas) {
-    const schema = s as string;
-    if (isGlob(schema) || isValidPath(schema)) {
-      files.push(schema);
-    }
-  }
+  const globalPatternSet = makeGlobalPatternSet(initialContext);
+  const localPatternSets = Object.keys(config.generates)
+    .map(filename => normalizeOutputParam(config.generates[filename]))
+    .map(conf => makeLocalPatternSet(conf));
+  const allPatterns = flattenPatternSets([globalPatternSet, ...localPatternSets]);
 
-  if (typeof config.watch !== 'boolean') {
-    files.push(...normalizeInstanceOrArray<string>(config.watch));
-  }
+  const shouldRebuild = makeShouldRebuild({ globalPatternSet, localPatternSets });
 
   let watcherSubscription: Awaited<ReturnType<typeof subscribe>>;
 
   const runWatcher = async () => {
-    const watchDirectory = await findHighestCommonDirectory(files);
+    const watchDirectory = await findHighestCommonDirectory(allPatterns);
```

Comment:
To ensure we keep the same behavior as from #9270, we need to flatten the list of patterns, which is the equivalent of what `files` was before (basically: all possible patterns). This will include a mix of negated and non-negated patterns, but it doesn't make any difference, because `mm.scan(pattern).base("!/foo/bar")` is `/foo/bar`.